client/order/epochqueue.go

```diff
@@ -0,0 +1,165 @@
+	_, ok := eq.orders[oid]
+	if ok {
+		return fmt.Errorf("%s is already queued", oid)
+	}
+
+	var commitment order.Commitment
+	copy(commitment[:], note.Commitment)
+
+	eq.mtx.Lock()
+	eq.orders[oid] = commitment
+	eq.mtx.Unlock()
+
+	return nil
+}
+
+// Size returns the number of entries in the epoch queue.
+func (eq *EpochQueue) Size() int {
+	eq.mtx.Lock()
+	size := len(eq.orders)
+	eq.mtx.Unlock()
+	return size
+}
+
+// Exists checks if the provided order id is in the queue.
+func (eq *EpochQueue) Exists(oid order.OrderID) bool {
+	eq.mtx.Lock()
+	_, ok := eq.orders[oid]
+	eq.mtx.Unlock()
+	return ok
+}
+
+// GenerateMatchProof calculates the sorting seed used in order matching
+// as well as the commitment checksum from the provided epoch queue
+// preimages and misses.
+//
+// The epoch queue needs to be reset if there are preimage mismatches or
+// non-existent orders for preimage errors.
+func (eq *EpochQueue) GenerateMatchProof(preimages []order.Preimage, misses []order.OrderID) (int64, msgjson.Bytes, error) {
+	eq.mtx.Lock()
+	defer eq.mtx.Unlock()
+
+	if len(eq.orders) == 0 {
+		return 0, nil,
+			fmt.Errorf("cannot generate match proof with an empty epoch queue")
+	}
+
+	// Remove all misses.
+	for _, oid := range misses {
+		delete(eq.orders, oid)
+	}
+
+	if len(eq.orders) != len(preimages) {
+		return 0, nil, fmt.Errorf("expected same length for epoch queue (%v) "+
+			"preimages (%v)", len(eq.orders), len(preimages))
+	}
+
+	// Map the preimages received with their associated epoch order ids.
+	matches := make(map[order.OrderID]order.Preimage, len(eq.orders))
+	for i := range preimages {
+		var match bool
+		for oid, commit := range eq.orders {
+			commitment := blake256.Sum256(preimages[i][:])
+			if bytes.Equal(commit[:], commitment[:]) {
+				matches[oid] = preimages[i]
+				match = true
+				break
+			}
+		}
+
+		if !match {
+			return 0, nil,
+				fmt.Errorf("no order match found for preimage %x", preimages[i])
+		}
+	}
+
+	// Ensure all remaining epoch orders matched to a preimage.
+	if len(matches) != len(eq.orders) {
+		return 0, nil, fmt.Errorf("expected all remaining epoch orders (%v) "+
+			"matched to a preimage (%v)", len(matches), len(eq.orders))
+	}
+
+	oids := make([]order.OrderID, 0, len(eq.orders))
+	commitments := make([]order.Commitment, 0, len(eq.orders))
+	for oid, commit := range eq.orders {
+		oids = append(oids, oid)
+		commitments = append(commitments, commit)
+	}
+
+	sort.Slice(oids, func(i, j int) bool {
+		return bytes.Compare(oids[i][:], oids[j][:]) < 0
+	})
+
+	sort.Slice(commitments, func(i, j int) bool {
+		return bytes.Compare(commitments[i][:], commitments[j][:]) < 0
+	})
+
+	// Concatenate the sorted preimages per and generate the
+	// seed.
+	sbuff := make([]byte, 0, len(eq.orders)*order.PreimageSize)
+	for _, oid := range oids {
+		pimg := matches[oid]
+		sbuff = append(sbuff, pimg[:]...)
+	}
+
+	sum := blake256.Sum256(sbuff)
+	seed := int64(binary.LittleEndian.Uint64(sum[:8]))
```

Comment:
Oh, I missed this.  The seed is the hash, not the first 8 bytes of it as an int64.  So return `seed`, `[]byte` is used to hold the "32-byte number" described by the spec.

Reply:
Ah, I need to update the spec as we used a 64-bit number before but a more appropriate seed is an array.  That really should have been updated when I did the mt19937 change, sorry.

Reply:
Wow, I thought I updated the spec for the shuffle change, but I did not.  On my todo list.